core/Pos/Binary/Crypto.hs

```diff
@@ -43,263 +34,192 @@ import           Pos.Crypto.Signing         (ProxyCert (..), ProxySecretKey (..)
-constantSizedBinaryToStoreGet :: Binary.Binary a => Int -> Store.Peek a
-constantSizedBinaryToStoreGet bytes = do
-    x <- getBytes bytes
-    case Binary.decodeOrFail (BSL.fromStrict x) of
-        Left (_, _, err) -> fail err
-        Right (bs, _, res)
-            | BSL.null bs -> pure res
-            | otherwise   -> fail "unconsumed input"
-
--- [CSL-1122] TODO: more efficient 'put' and 'get' are possible if 'Store'
--- instances are added into pvss-haskell
-#define BiPvss(T, PT, BYTES) \
+#define BiPvss(T, PT) \
   instance Bi T where {\
-    size = ConstSize BYTES ;\
-    put = labelP "T" . putBytes . BSL.toStrict . Binary.encode ;\
-    get = label "T" $ constantSizedBinaryToStoreGet BYTES };\
+    encode = encodeBinary ;\
+    decode = decodeBinary };\
   deriving instance Bi PT ;\
 
-BiPvss (Pvss.PublicKey, VssPublicKey, 33)    -- yes it's 33 and not 32
-BiPvss (Pvss.KeyPair, VssKeyPair, 65)        -- 32+33
-BiPvss (Pvss.Secret, Secret, 33)
-BiPvss (Pvss.DecryptedShare, Share, 101)     -- 4+33+64
-BiPvss (Pvss.EncryptedShare, EncShare, 101)
-BiPvss (Pvss.Proof, SecretProof, 64)
-
-instance Store.Store Pvss.ExtraGen where
-    size = ConstSize 33
-    poke = labelP "Pvss.ExtraGen" . putBytes . BSL.toStrict . Binary.encode
-    peek = label "Pvss.ExtraGen" $ constantSizedBinaryToStoreGet 33
-instance Store.Store Pvss.Commitment where
-    size = ConstSize 33
-    poke = labelP "Pvss.Commitment" . putBytes . BSL.toStrict . Binary.encode
-    peek = label "Pvss.Commitment" $ constantSizedBinaryToStoreGet 33
-
-Store.makeStore ''SecretSharingExtra
+BiPvss (Pvss.PublicKey, VssPublicKey)
+BiPvss (Pvss.KeyPair, VssKeyPair)
+BiPvss (Pvss.Secret, Secret)
+BiPvss (Pvss.DecryptedShare, Share)
+BiPvss (Pvss.EncryptedShare, EncShare)
+BiPvss (Pvss.Proof, SecretProof)
+
+instance Bi Pvss.ExtraGen where
+    encode = encodeBinary
+    decode = decodeBinary
+
+instance Bi Pvss.Commitment where
+    encode = encodeBinary
+    decode = decodeBinary
+
 instance Bi SecretSharingExtra where
-    put = labelP "SecretSharingExtra" . Store.poke
-    get = label "SecretSharingExtra" $ Store.peek
-    size = Store.size
+    encode (SecretSharingExtra eg comms) = encodeListLen 2
+                                        <> encode eg
+                                        <> encode comms
+    decode = SecretSharingExtra
+          <$  enforceSize "SecretSharingExtra" 2
+          <*> decode
+          <*> decode
 
 deriving instance Bi (AsBinary SecretSharingExtra)
 
 ----------------------------------------------------------------------------
 -- SecretSharing AsBinary
 ----------------------------------------------------------------------------
 
-#define BiMacro(B, BYTES) \
+#define BiMacro(B) \
   instance Bi (AsBinary B) where {\
-    size = ConstSize BYTES ;\
-    put (AsBinary bs) = putBytes bs ;\
-    get = label "B (BYTES bytes)" $ AsBinary <$> getBytes BYTES}; \
+    encode (AsBinary bs) = encode bs ;\
+    decode = AsBinary <$> decode}; \
```

Comment:
Ok, makes sense, created https://issues.serokell.io/issue/CSL-1412 and https://issues.serokell.io/issue/CSL-1413 to keep track.

> So, if the input string is fixed, also the overhead CBOR introduces is predictable, so I think for now everything should continue working

Yes, with regards to this it should work. I just don't like unnecessary overhead. The way we store `AsBinary VssPublicKey` now is that we wrap `VssPublicKey` into CBOR bytestring once in instance for `VssPublicKey` and then one more time in instance for `AsBinary`. One wrapping is unavoidable (it's enforced by CBOR), but another one is not necessary in principle. And previously there was no such wrapping, because we could put fixed-size string directly. The idea of `AsBinary` type is that it shouldn't add overhead.

To sum it up, there are two issues, one of them is about a serious bug (CSL-1412), another one about small inefficiency (CSL-1413). The former must be fixed and I assigned it `Critical` priority and set `Required for mainnet` tag. The latter is not so important (I personally don't like unnecessary overhead, especially size overhead, because for blockchain size matters, but I admit that the overhead is very small), maybe we don't need to do it, I assigned it `Normal` priority.